### PR TITLE
feat: add React/Next.js-style optimistic() state API to @webjsdev/core

### LINKED ISSUE #798: feat: add React/Next.js-style optimistic() state API to @webjsdev/core
# Feature Request: Introduce React/Next.js-style `optimistic()` State API in `@webjsdev/core`

## 1. Goal & Context
Currently, performing Snappy UX / Optimistic UI updates in `webjs` Web Components requires developers to manually duplicate and mutate their local reactive properties (e.g. `this.todos`), track temporary IDs, implement try-catch logic, and manually rollback the array to a cached copy if the network request fails. 

This proposal aims to introduce a declarative helper named `optimistic()` under `@webjsdev/core` with full parity to React 19 / Next.js's `useOptimistic` hook, but customized for class-based custom elements and reactive signals.

---

## 2. API Design & React 19 Feature Parity
The signature of React 19's `useOptimistic` is:
```typescript
function useOptimistic<State, Action>(
  passthrough: State,
  reducer?: (state: State, action: Action) => State
): [State, (action: Action) => void];
```
To achieve complete feature parity, our `optimistic()` helper will implement:
1. **Optional Reducer**: If the `update` (reducer) function is omitted, it defaults to simply replacing/setting the state to the optimistic payload.
2. **Auto-Promise Release (Next.js Action Parity)**: React's optimistic state resets automatically when the enclosing Transition/Action promise resolves. In `webjs`, we support this cleanly by allowing `.add(payload, promise)` to take an optional Promise as the second argument. When passed, the wrapper automatically releases the optimistic state when the promise resolves (via `.finally()`), eliminating the `try-finally` boilerplate for the developer.

### Developer Interface
```ts
import { WebComponent, prop, optimistic, html } from '@webjsdev/core';
import { createTodo } from '../modules/todos/actions/create-todo.server.ts';

export class TodoList extends WebComponent({
  todos: prop<Todo[]>(Array),
}) {
  // 1. Declare the wrapper (with optional reducer)
  private optimisticTodos = optimistic(this, {
    source: () => this.todos,
    update: (state, newTitle: string) => [
      ...state,
      { id: 'temp', title: newTitle, completed: false, pending: true }
    ],
  });

  constructor() {
    super();
    this.todos = [];
  }

  async handleSubmit(e: SubmitEvent) {
    e.preventDefault();
    const title = 'New Task';

    // 2. Add with a Promise for auto-release (no try-finally or release() call required!)
    const actionPromise = createTodo({ title });
    this.optimisticTodos.add(title, actionPromise);

    const result = await actionPromise;
    if (result.success && result.data) {
      // 3. Update the source of truth
      this.todos = [...this.todos, result.data];
    }
  }

  render() {
    return html`
      <ul>
        ${this.optimisticTodos.value.map(todo => html`
          <li class=${todo.pending ? 'opacity-50' : ''}>${todo.title}</li>
        `)}
      </ul>
    `;
  }
}
```

---

## 3. Internal Implementation Specification

```ts
export function optimistic<State, Action>(
  host: ReactiveInterface,
  options: {
    source: () => State;
    update?: (state: State, action: Action) => State;
  }
): OptimisticState<State, Action>;
```

### Class Invariants & Internal Logic:
1. **Host Connection**: It must register as a Reactive Controller or use the host's lifecycle trigger (`host.requestUpdate()`) to schedule a component re-render when the optimistic state changes.
2. **Queued Updates**: It must maintain an internal array of active optimistic updates:
   ```ts
   private updates: { id: string; payload: Action }[] = [];
   ```
3. **Computed `.value` Getter**: The `.value` property must compute the state dynamically:
   - Read the current value from the `source()` function.
   - Fold/reduce all active items in the `updates` queue. If no `update` reducer is provided, simply use the last payload.
   ```ts
   get value(): State {
     let current = this.options.source();
     if (!this.options.update) {
       return this.updates.length > 0 
         ? (this.updates[this.updates.length - 1].payload as any)
         : current;
     }
     for (const update of this.updates) {
       current = this.options.update(current, update.payload);
     }
     return current;
   }
   ```
4. **`.add(payload, promise?)` Trigger**:
   - Generate a unique ID.
   - Push `{ id, payload }` to the `updates` array.
   - Call `host.requestUpdate()` to update the DOM immediately.
   - Construct the `release` callback:
     ```ts
     const release = () => {
       this.updates = this.updates.filter(u => u.id !== id);
       this.host.requestUpdate();
     };
     ```
   - If an optional `promise` is provided, chain `.finally(() => release())` to it.
   - Return the `release` function.

---

## 4. Verification Plan
- **Unit Tests** (`packages/core/test/state/optimistic.test.js`):
  - Verify default behavior when `update` reducer is omitted (sets value directly).
  - Verify manual `release()` callback flow.
  - Verify auto-release flow when an async action Promise is passed to `.add()`.
  - Verify concurrent updates chain correctly.

### PR DESCRIPTION
**Pull request by vivek7405**
Closes #798

Introduces a React 19 / Next.js-style declarative `optimistic()` state wrapper for Web Components, with backward compatibility for the existing signal-based imperative API.

## Problem

Performing optimistic UI updates in webjs Web Components currently requires developers to manually duplicate and mutate local reactive properties, track temporary IDs, implement try-catch logic, and manually roll back to a cached copy if the network request fails. This is verbose, error-prone, and does not match the ergonomics of React 19's `useOptimistic` hook.

## What this does

- **`OptimisticState` class**: A new declarative wrapper that tracks a queue of pending optimistic updates and computes the combined value via a reducer. Created via `optimistic(host, { source, update })`.
- **`.add(payload, promise?)`**: Pushes an optimistic update and schedules a re-render. When a `promise` is provided, auto-releases the update on settlement (via `.finally()` or a `.then()` fallback for thenables lacking `.finally`), eliminating try-finally boilerplate.
- **Optional reducer**: When `update` is omitted, the payload replaces the state directly (`Action = State`), matching the simple `useOptimistic(setState)` pattern.
- **Concurrent updates**: Multiple `.add()` calls stack; each `release()` removes its own update by ID, so concurrent optimistic operations resolve independently.
- **Backward compatibility**: The existing `optimistic(signal, value, action)` signal-based imperative API is preserved unchanged. The exported `optimistic()` function dispatches by detecting a signal-like object (`get`/`set` methods) vs. a host/options pair.
- **TypeScript overloads**: Two separate overloads for the declarative signature (with and without `update`) so `Action` infers correctly without explicit type annotations.

## Tests

- `packages/core/test/signals/optimistic-declarative.test.js` (8 unit tests: initial value, manual add/release, default reducer, auto-release on resolve, auto-release on reject, concurrent updates, host without `requestUpdate`, thenable without `finally`)
- `packages/core/test/signals/optimistic.test.js` (6 existing legacy tests, all passing)
- `packages/core/test/signals/browser/optimistic.test.js` (3 browser tests for the signal-based API with real WebComponent rendering, plus 1 new declarative browser test)
- Bun matrix: all pass (the new `node:test` files run under Bun too)

## Docs synced

- `agent-docs/advanced.md`: Full reference with both declarative and imperative signatures
- `agent-docs/recipes.md`: Updated optimistic todo recipe to use declarative API as preferred pattern
- `docs/app/docs/client-router/page.ts`: Docs site section updated with both signatures
- `AGENTS.md`: Public API table updated
- `blog/optimistic-ui-without-boilerplate.md`: SEO-optimized blog post

## Scope

- New public API: `optimistic(host, options)` returns `OptimisticState<State, Action>` with `.value` getter and `.add(payload, promise?)` method.
- No breaking changes to the existing `optimistic(signal, value, action)` imperative API.
- The `optimistic` import remains classified as a reactive (client-work) import in the elision analyser, so a component using it is never elided as display-only.
- UI package (Tier 1 + Tier 2 components): no changes needed. These are presentational primitives that do not manage mutation state.
- Website and ui website: no changes needed. No optimistic UI references exist in marketing copy.

### CODE REVIEW

File: packages/core/src/optimistic.js [hunk 1]
```diff
@@ -1,64 +1,81 @@
...
+    for (const update of this.updates) {
+      current = this.options.update(current, update.payload);
+    }
+    return current;
+  }
+
+  add(payload, promise) {
+    const id = `opt-${++this._nextId}`;
+    this.updates.push({ id, payload });
+    this.host.requestUpdate();
```
Comment (vivek7405) [inline packages/core/src/optimistic.js]: Calling `this.host.requestUpdate()` directly might throw if the host is missing or does not implement the method. Consider guarding it: `if (this.host && typeof this.host.requestUpdate === 'function') { this.host.requestUpdate(); }`.

**Review by vivek7405 — COMMENTED** [review]
### PR Review Summary

I have reviewed the declarative optimistic state API PR. The implementation is clean and integrates nicely with the existing codebase. Here is a summary of findings and suggestions:

1. **Naming & Cleanliness:** Naming is consistent. The overloaded function parameters (`first, second, third`) in JavaScript are generic but acceptable for pattern-matching overloads.
2. **Type Overloading:** The current TypeScript type definition for the declarative signature has an optional `update` parameter. When `update` is omitted, TypeScript fails to infer `Action`, defaulting it to `unknown` and forcing the caller to explicitly type it. Splitting this into two separate overloads (one with `update` and one without where `Action = State`) makes the API much more ergonomic.
3. **Test Coverage & Edge Cases:** Verified that the existing test cases pass. However, there are two edge cases not covered by tests that can cause runtime exceptions:
   - Custom thenable objects passed to `add()` that lack a `.finally` method.
   - Mock/custom hosts passed to `optimistic()` that lack a `requestUpdate` method.
4. **Runtime Bugs & Unhandled Rejections:**
   - Calling `promise.finally()` directly can crash on custom thenables lacking a `.finally` method with `TypeError: promise.finally is not a function`.
   - Calling `this.host.requestUpdate()` without a safety check can crash if the host does not implement it.

File: packages/core/src/optimistic.js [hunk 2]
```diff
@@ -1,64 +1,81 @@
...
+
+    const release = () => {
+      const idx = this.updates.findIndex(u => u.id === id);
+      if (idx !== -1) {
+        this.updates.splice(idx, 1);
+        this.host.requestUpdate();
```
Comment (vivek7405) [inline packages/core/src/optimistic.js]: Same as above, guard `this.host.requestUpdate()` inside the release handler to prevent TypeErrors on hosts lacking the method.

File: packages/core/src/optimistic.js [hunk 3]
```diff
@@ -1,64 +1,81 @@
...
+      }
+    };
+
+    if (promise && typeof promise.then === 'function') {
+      promise.finally(() => release()).catch(() => {});
```
Comment (vivek7405) [inline packages/core/src/optimistic.js]: Custom thenables (promise-like objects) may have `.then` but lack a `.finally` method. Check if `.finally` exists on the promise first, and fall back to `.then` to prevent `TypeError: promise.finally is not a function`:

```javascript
if (typeof promise.finally === 'function') {
  promise.finally(() => release()).catch(() => {});
} else {
  promise.then(() => release(), () => release());
}
```

File: packages/core/index.d.ts [hunk 4]
```diff
@@ -113,11 +113,24 @@ export { stringify, parse, serialize, deserialize } from './src/serialize.js';
...
-// Optimistic-mutation helper: set a signal to an expected value immediately,
-// run the action, roll back on a thrown error or a `{ success: false }`
-// ActionResult, keep the value on success. Returns the action's result.
+export interface OptimisticState<State, Action> {
+  readonly value: State;
+  add(payload: Action, promise?: Promise<any> | any): () => void;
+}
+
+// Declarative Signature (React 19 / Next.js inspired)
+export function optimistic<State, Action>(
```
Comment (vivek7405) [inline packages/core/index.d.ts:122]: Since `update` is optional, if a user omits it (falling back to the default reducer), TypeScript will not be able to infer the `Action` parameter and will default it to `unknown` (requiring manual type annotation).

Consider splitting this into two overloaded signatures to auto-infer `Action = State` when `update` is omitted:

```typescript
// Declarative Signature - with update function
export function optimistic<State, Action>(
  host: { requestUpdate(): void },
  options: {
    source: () => State;
    update: (state: State, action: Action) => State;
  }
): OptimisticState<State, Action>;

// Declarative Signature - without update function
export function optimistic<State>(
  host: { requestUpdate(): void },
  options: {
    source: () => State;
  }
): OptimisticState<State, State>;
```
